pkg/scheduler/algorithm_factory.go

```diff
@@ -400,15 +400,36 @@ func RegisterCustomPriorityFunction(policy schedulerapi.PriorityPolicy, args *pl
 				Weight: policy.Weight,
 			}
 		} else if policy.Argument.LabelPreference != nil {
-			pcf = &PriorityConfigFactory{
-				MapReduceFunction: func(args PluginFactoryArgs) (priorities.PriorityMapFunction, priorities.PriorityReduceFunction) {
-					return priorities.NewNodeLabelPriority(
-						policy.Argument.LabelPreference.Label,
-						policy.Argument.LabelPreference.Presence,
-					)
-				},
-				Weight: policy.Weight,
+			// We use the NodeLabel plugin name for all NodeLabel custom priorities.
+			// It may get called multiple times but we essentially only register one instance of NodeLabel priority.
+			// This name is then used to find the registered plugin and run the plugin instead of the priority.
+			name = nodelabel.Name
+			if args.NodeLabelArgs == nil {
+				args.NodeLabelArgs = &nodelabel.Args{}
+			}
+			if policy.Argument.LabelPreference.Presence {
+				args.NodeLabelArgs.PresentLabelsPreference = append(args.NodeLabelArgs.PresentLabelsPreference, policy.Argument.LabelPreference.Label)
+			} else {
+				args.NodeLabelArgs.AbsentLabelsPreference = append(args.NodeLabelArgs.AbsentLabelsPreference, policy.Argument.LabelPreference.Label)
+			}
+			schedulerFactoryMutex.RLock()
+			if existing, ok := priorityFunctionMap[name]; ok {
+				// If there are n NodeLabel priority configured in the policy, the weight for the corresponding
+				// priority is n*(weight of each priority in policy).
+				existing.Weight += policy.Weight
+				pcf = &existing
+			} else {
+				pcf = &PriorityConfigFactory{
+					MapReduceFunction: func(_ PluginFactoryArgs) (priorities.PriorityMapFunction, priorities.PriorityReduceFunction) {
+						return priorities.NewNodeLabelPriority(
+							args.NodeLabelArgs.PresentLabelsPreference,
+							args.NodeLabelArgs.AbsentLabelsPreference,
+						)
+					},
+					Weight: policy.Weight,
+				}
```

Comment:
we can track the weight in NodeLabelArgs, and replace the above code with:

```
pcf = &PriorityConfigFactory{
  MapReduceFunction: func(_ PluginFactoryArgs) (priorities.PriorityMapFunction, priorities.PriorityReduceFunction) {
  return priorities.NewNodeLabelPriority(
    args.NodeLabelArgs.PresentLabelsPreference,
    args.NodeLabelArgs.AbsentLabelsPreference,
  )
},
Weight: args.NodeLabelArgs.Weight,
}
```

Reply:
Discussed offline, we will still use this approach to accumulate the weight otherwise we have two different source of weight for plugins.

However there is a bug in the current implementation that the labels are not updated properly. I fixed it.